File: problems/Reverse Integer/reverse.java
class Solution {
    public int reverse(int x) {
        int result = 0;

        while (0 != x) {
            int digit = x % 10;
            x /= 10;

            if (willBeOverflow(result, digit)) {
                return 0;
            }

            result = 10 * result + digit;
        }

        return result;
    }

    private boolean willBeOverflow(int num, int digit) {
        if (Integer.MAX_VALUE / 10 < num || (Integer.MAX_VALUE / 10 == num && 7 < digit)) {
            return true;
        }

        if (Integer.MIN_VALUE / 10 > num || (Integer.MIN_VALUE / 10 == num && -8 > digit)) {
            return true;
        }

        return false;
    }
}

File: problems/Reverse Integer/reverse.js
/**
 * @param {number} x
 * @return {number}
 */
var reverse = function(x) {
    let result = 0;
    const MAX_VALUE = Math.pow(2, 31) - 1;
    const MIN_VALUE = Math.pow(-2, 31);

    while (x) {
        const digit = x % 10;
        x = Math.trunc(x/10);

        if (Math.trunc(MAX_VALUE/10) < result ||
            (Math.trunc(MAX_VALUE/10) === result && digit > 7)) {
            return 0;
        }

        if (Math.trunc(MIN_VALUE/10) > result ||
            (Math.trunc(MIN_VALUE/10) === result && digit < -8)) {
            return 0;
        }

        result *= 10;
        result += digit;
    }

    return result;
};
